### PR TITLE
*: support MySQL backend

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Apply simplest change to support the MySQL backend.

### What is changed and how it works?

* Added the `[tikv-importer] backend` task configuration and `--backend` command line option.
* When `backend` is `"mysql"`, choose the new MySQL backend (when `backend` is "importer" choose the normal Importer backend)
* The MySQL backend reuses the TiDB SQL connection. Therefore, we include the SQL mode when opening the connection. 
* The MySQL backend simply reconstructs a multi-valued INSERT statement. Prepared statements (`INSERT VALUES (?,?,?)`) is not used due to complication with hex/binary literals.

Caveats:

* I have only enabled the MySQL backend on two integration tests for now. In the *next PR* I'll modify the integration test procedure so every test case will be tested against both Import and MySQL backends.
* The current implementation is not tested for speed yet and *may* be slower than Loader. 

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

Related changes

 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: lightning/config/config.go [hunk 1]
```diff
@@ -290,6 +295,13 @@ func (cfg *Config) Adjust() error {
 		}
 	}
 
+	cfg.TikvImporter.Backend = strings.ToLower(cfg.TikvImporter.Backend)
+	switch cfg.TikvImporter.Backend {
+	case "mysql", "importer":
+	default:
+		return errors.New("invalid config: unsupported `tikv-importer.backend`")
```
Comment (csuzhangxc) [inline lightning/config/config.go]: including the value of `cfg.TikvImporter.Backend` in the error message?

Comment (lance6716) [reply lightning/config/config.go]: seems that those errors from this function not including original value.

File: tidb-lightning.toml [hunk 2]
```diff
@@ -61,6 +61,9 @@ driver = "file"
 #keep-after-success = false
 
 [tikv-importer]
+# Delivery backend, can be "importer" or "mysql".
+#backend = "importer"
```
Comment (lance6716) [inline tidb-lightning.toml]: shouldn't comment methinks

File: lightning/restore/tidb.go [hunk 3]
```diff
@@ -44,7 +44,7 @@ type TiDBManager struct {
 }
 
 func NewTiDBManager(dsn config.DBStore) (*TiDBManager, error) {
-	db, err := common.ConnectDB(dsn.Host, dsn.Port, dsn.User, dsn.Psw)
+	db, err := common.ConnectDB(dsn.Host, dsn.Port, dsn.User, dsn.Psw, dsn.StrSQLMode)
```
Comment (lance6716) [inline lightning/restore/tidb.go]: > The MySQL backend reuses the TiDB SQL connection. Therefore, we include the SQL mode when opening the connection.

Now things are getting mixed up 😛

File: mock/backend.go [hunk 4]
```diff
@@ -149,6 +149,18 @@ func (mr *MockBackendMockRecorder) RetryImportDelay() *gomock.Call {
 	return mr.mock.ctrl.RecordCallWithMethodType(mr.mock, "RetryImportDelay", reflect.TypeOf((*MockBackend)(nil).RetryImportDelay))
 }
 
+// ShouldPostProcess mocks base method
+func (m *MockBackend) ShouldPostProcess() bool {
+	ret := m.ctrl.Call(m, "ShouldPostProcess")
```
Comment (lance6716) [inline mock/backend.go]: I don't understand the purpose of this function, is it calling itself?

File: tidb-lightning.toml (hunk 2 above)
Comment (ericsyh) [reply tidb-lightning.toml]: +1

File: mock/backend.go (hunk 4 above)
Comment (kennytm) [reply mock/backend.go]: This file is generated by [mockgen](https://github.com/golang/mock), no need to review its content 😅

**Comment by GregoryIan** [comment]
Good Job!

File: lightning/kv/mysql.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kv
```
Comment (GregoryIan) [inline lightning/kv/mysql.go]: package name `backend` is better

File: lightning/kv/mysql.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+		sb.WriteString("NULL")
+
+	case types.KindMinNotNull:
+		sb.WriteString("MINVALUE")
+
+	case types.KindMaxValue:
+		sb.WriteString("MAXVALUE")
+
+	case types.KindInt64:
+		var buffer [20]byte
```
Comment (GregoryIan) [inline lightning/kv/mysql.go]: why `20`？

File: lightning/kv/mysql.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+func (enc mysqlEncoder) Encode(_ log.Logger, row []types.Datum, _ int64, _ []int) (Row, error) {
+	var encoded strings.Builder
+	encoded.Grow(8 * len(row))
+	encoded.WriteByte('(')
+	for i, field := range row {
+		if i != 0 {
+			encoded.WriteByte(',')
+		}
+		if err := enc.appendSQL(&encoded, &field); err != nil {
+			return nil, err
```
Comment (GregoryIan) [inline lightning/kv/mysql.go]: return more information with error, like which column or column index?

**Comment by GregoryIan** [comment]
Rest LGTM

**Comment by GregoryIan** [comment]
@kennytm  we can merge it ASAP after resolving these comments,

**Comment by lance6716** [comment]
LGTM

**Comment by kennytm** [comment]
/run-all-tests

(BTW zhouqiang tested this yesterday and found that the speed is 3x slower than Loader. The next step is to use pprof and find out why it's so slow)

**Comment by GregoryIan** [comment]
agree with you @kennytm (but @zhouqiang-cl tell me that the speed is 10x slower than Loader)

**Comment by kennytm** [comment]
/run-all-tests

```
19:44:53  GO111MODULE=off go get github.com/mattn/goveralls
19:46:15  Bad response status from coveralls: 405
19:46:15  <html>
19:46:15  <head><title>405 Not Allowed</title></head>
19:46:15  <body bgcolor="white">
19:46:15  <center><h1>405 Not Allowed</h1></center>
19:46:15  <hr><center>nginx</center>
19:46:15  </body>
19:46:15  </html>
```

**Comment by lance6716** [comment]
seems coveralls is [under maintenance](https://status.coveralls.io/?utm_source=embed)

**Comment by kennytm** [comment]
😢

**Comment by kennytm** [comment]
/run-all-tests

>  Update - Database is back online, processing queue is working through backlog. 
>
>  Aug 13, 21:22 PDT

**Pull request merged**